src/Init/Data/Range/Polymorphic/UInt.lean

```diff
@@ -0,0 +1,344 @@
+/-
+Copyright (c) 2025 Lean FRO, LLC. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Paul Reichert
+-/
+module
+
+prelude
+public import Init.Data.Range.Polymorphic.Instances
+public import Init.Data.Order.Lemmas
+public import Init.Data.UInt
+import Init.Omega
+
+public section
+
+open Std Std.PRange
+
+namespace UInt8
+
+instance : UpwardEnumerable UInt8 where
+  succ? i := if i + 1 = 0 then none else some (i + 1)
+  succMany? n i := if h : i.toNat + n < UInt8.size then some (.ofNatLT _ h) else none
+
+instance : LawfulUpwardEnumerable UInt8 where
+  ne_of_lt a b h := by
+    obtain ⟨c, hc⟩ := h
+    simp [UpwardEnumerable.succMany?, ← UInt8.toNat_inj, UInt8.size] at hc ⊢
+    omega
+  succMany?_zero x := by simp [UpwardEnumerable.succMany?, x.toNat_lt_size]
+  succMany?_succ a b := by
+    simp only [UpwardEnumerable.succMany?, UInt8.size, UpwardEnumerable.succ?]
+    split <;> split <;> simp [← UInt8.toNat_inj] <;> omega
+
+instance : LawfulUpwardEnumerableLE UInt8 where
+  le_iff x y := by
+    simp [UpwardEnumerable.LE, UpwardEnumerable.succMany?, UInt8.le_iff_toNat_le]
+    apply Iff.intro
+    · intro hle
+      refine ⟨y.toNat - x.toNat, ?_⟩
+      apply Exists.intro
+      · rw [ofNatLT_sub (UInt8.toNat_lt y) hle, ofNatLT_toNat, ofNatLT_toNat, UInt8.add_comm,
+          UInt8.sub_add_cancel]
+      · rw [Nat.add_sub_cancel' ‹_›]
+        exact UInt8.toNat_lt _
+    · rintro ⟨n, hn, rfl⟩
+      rw [UInt8.toNat_add, UInt8.toNat_ofNatLT, Nat.mod_eq_of_lt hn]
+      omega
```

Comment:
Rather than repeating this for each type, I wonder if you can prove it for bitvec, and then pull back over a theorem like `(succ? i).toBitVec = succ? i.toBitVec`.

Reply:
(you of course still want to define the data-carrying instances separately to get the optimal C implementations)

Reply:
Good idea, I've added `BitVec` support now.